src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -60,21 +60,29 @@ def __init__(
     def resolve(self, root_reqs, check_supported_wheels):
         # type: (List[InstallRequirement], bool) -> RequirementSet
 
-        # FIXME: Implement constraints.
-        if any(r.constraint for r in root_reqs):
-            raise InstallationError("Constraints are not yet supported.")
-
         provider = PipProvider(
             factory=self.factory,
             ignore_dependencies=self.ignore_dependencies,
         )
         reporter = BaseReporter()
         resolver = RLResolver(provider, reporter)
 
-        requirements = [
-            self.factory.make_requirement_from_install_req(r)
-            for r in root_reqs
-        ]
+        # Build the list of root requirements, making sure that
```

Comment:
Does this imply that constraints are only used for the root requirements, and not secondary requirements?

Reply:
No, constraints only come *into* pip as root requirements (i.e. from a constraints file). They never appear as dependencies - dependencies are always "full" requirements, which will be installed.

Constraints are applied along with normal requirements to determine what are the allowable versions for any project.

Reply:
OK, I think that makes sense.